deku-derive/src/macros/deku_read.rs

```diff
@@ -352,6 +364,24 @@ fn emit_enum(input: &DekuData) -> Result<TokenStream, syn::Error> {
         });
     }
 
+    // create `deku_id()` function for retrieving at runtime the id value
+    let deku_id_id_type = if let Some(id_type) = id_type {
+        id_type
+    } else {
+        &ctx_types
```

Comment:
The problem with using `ctx_types` is when there's more than one, it breaks:

```rust
#[derive(PartialEq, Debug, DekuRead, DekuWrite)]
#[deku(ctx = "test: u16, my_id: u8", id = "my_id")]
enum EnumId {
    #[deku(id = "1")]
    VarA(u8),
    #[deku(id = "2")]
    VarB,
}
```

generates

```rust
impl DekuEnumExt<'_, (u16, u8)> for EnumId {
    fn deku_id(&self) -> Result<(u16, u8), DekuError> {
        match self {
            Self::VarA(__deku_field_0) => Ok(1),
            Self::VarB => Ok(2),
            _ => Err(DekuError::IdVariantNotFound),
        }
    }
}
```

One way of solving this would be to find the type of `input.id` in `input.ctx`

#[deku(ctx = "test: u16, **my_id**: **u8**", id = "**my_id**")]

Having a utility function in `macros/mod.rs` would be useful. You can look at `gen_ctx_types_and_arg` to see how it iterates `ctx`

Reply:
Just pushed https://github.com/sharksforarms/deku/pull/176/commits/649b1d323621fcae5367208fb78c434cee3d0dd0 with this fix. It works and the tests compile but the code quality might be lacking, any opinions or extra tests are good ;)